File: Solutions/0318/0318.js
var maxProduct = function(words) {
    let ans = 0;

    const getMask = (word) => {
      let mask = 0;
      for (const c of word) {
        mask |= 1 << (c.charCodeAt(0) - 'a'.charCodeAt(0));
      }
      return mask;
    };

    const masks = words.map(getMask);

    for (let i = 0; i < words.length; i++) {
      for (let j = 0; j < i; j++) {
        if (!(masks[i] & masks[j])) {
          ans = Math.max(ans, words[i].length * words[j].length);
        }
      }
    }

    return ans;
};

File: Solutions/0318/0318.go
func maxProduct(words []string) int {
	ans := 0

	getMask := func(word string) int {
		mask := 0
		for _, c := range word {
			mask |= 1 << (c - 'a')
		}
		return mask
	}

	masks := make([]int, len(words))
	for i, word := range words {
		masks[i] = getMask(word)
	}

	for i := 0; i < len(words); i++ {
		for j := 0; j < i; j++ {
			if masks[i]&masks[j] == 0 {
				ans = max(ans, len(words[i])*len(words[j]))
			}
		}
	}

	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
    